File: go/markov.go
package main

import "fmt"
import "io"
import "os"
import "math/rand"

const (
	NPREF  int = 2     // number of prefix words
	MAXGEN int = 10000 // maximum words generated
)

var statetab map[string][]string

var NOWWORD = "\n"

func arrayToString(prefix []string) string {
	s := ""
	for i := 0; i < len(prefix); i++ {
		s += prefix[i]
	}
	return s
}

func add(prefix []string, suffix string) []string {
	key := arrayToString(prefix)
	statetab[key] = append(statetab[key], suffix)

	prefix = prefix[1:NPREF]
	prefix = append(prefix, suffix)
	return prefix
}

func build(prefix []string) {
	var suffix string
	for n, err := fmt.Scan(&suffix); n != 0 && err != io.EOF; n, err = fmt.Scan(&suffix) {
		if err != nil {
			fmt.Println("error: ", err)
			os.Exit(1)
		}
		prefix = add(prefix, suffix)
	}
	add(prefix[0:NPREF], NOWWORD)
}

func generate(prefix []string) {
	result := ""

	for i := 0; i < NPREF; i++ {
		prefix[i] = NOWWORD
	}

	for i := 0; i < MAXGEN; i++ {
		suffixArr := statetab[arrayToString(prefix)]
		word := suffixArr[rand.Intn(len(suffixArr))]
		if word == NOWWORD {
			break
		}
		result += word + " "

		prefix = prefix[1:NPREF]
		prefix = append(prefix, word)
	}

	fmt.Println(result)
}

func main() {
	statetab = make(map[string][]string)

	var prefix [NPREF]string

	for i := 0; i < NPREF; i++ {
		prefix[i] = NOWWORD
	}
	build(prefix[0:NPREF])
	generate(prefix[0:NPREF])
}

File: javascript/markov.js
'use strict';

const fs = require('fs');

const NPREF = 2;
const MAXGEN = 10000;
const NOWWORD = '\n';

const statetab = {};
const prefix = [];

let result = '';

function add(prefix, suffix) {
    if (!statetab[prefix]) {
        statetab[prefix] = [suffix];
    } else {
        statetab[prefix].push(suffix);
    }
    prefix.shift();
    prefix.push(suffix);
}

function build(prefix, text) {
    text.forEach(word => {
        add(prefix, word);
    });
}

function generate() {
    for (let i = 0; i < NPREF; i++) {
        prefix[i] = NOWWORD;
    }

    for (let i = 0; i < MAXGEN; i++) {
        const suffix = statetab[prefix];
        const word = suffix[Math.floor(Math.random() * suffix.length)];
        if (word === NOWWORD)
            break;
        result += word + ' ';
        prefix.shift();
        prefix.push(word);
    }
}

fs.readFile('markov-chain-test-text.txt', (err, data) => {
    if (err) throw err;

    const text = data.toString().split(/[\s\n]/g).filter((word) => {
        return word.length > 0;
    });

    for (let i = 0; i < NPREF; i++) {
        prefix[i] = NOWWORD;
    }
    build(prefix, text);
    add(prefix, NOWWORD);
    generate();
    console.log(result);
});
